File: go/cote/LTCD/빗물_트래핑.go
package LTCD

import (
	"container/list"
	"fmt"
	"math/rand"
)

// const (
// 	TEST_CASE = 10
// 	N = 10
// 	MAX_HEIGHT = 100
// )

type 빗물_트래핑 struct {
	TEST_CASE  int
	N          int
	MAX_HEIGHT int
}

func (problem 빗물_트래핑) max(a, b int) int {
	if a < b {
		return b
	}
	return a
}

func (problem 빗물_트래핑) min(a, b int) int {
	if a < b {
		return a
	}
	return b
}

func (problem 빗물_트래핑) solve1(heights []int) int {
	volumne := 0
	left, right := int(0), len(heights)-1
	leftmax, rightmax := int(0), int(0)

	for ok := true; ok; ok = (left < right) {

		leftmax = problem.max(leftmax, heights[left])
		rightmax = problem.max(rightmax, heights[right])

		if leftmax <= rightmax {
			volumne += leftmax - heights[left]
			left++
		} else {
			volumne += rightmax - heights[right]
			right--
		}
	}

	return volumne
}

func (problem 빗물_트래핑) solve2(heights []int) int {

	volume := 0
	stack := list.New()

	for i, _ := range heights {

		for {
			if stack.Len() == 0 {
				break
			}

			top := stack.Back().Value.(int)

			if heights[i] <= heights[top] {
				break
			}

			cur := stack.Remove(stack.Back()).(int)

			if stack.Len() == 0 {
				break
			}

			left := stack.Back().Value.(int)
			distance := i - left - 1
			waters := min(heights[left], heights[i]) - heights[cur]
			volume += distance * waters

		}

		stack.PushBack(i)
	}

	return volume
}

func (problem 빗물_트래핑) main() {
	rand.Seed(41)

	for it := 0; it < problem.TEST_CASE; it++ {
		n := rand.Intn(problem.N-2) + 2
		h := rand.Intn(problem.MAX_HEIGHT-1) + 1

		heights := make([]int, n)

		for i, _ := range heights {
			heights[i] = rand.Intn(h)
		}

		fmt.Println(problem.solve1(heights), problem.solve2(heights))

	}
}

File: python/cote/LTCD/빗물_트래핑.py
from typing import *
import random

TEST_CASES = 10
N = 10
MAX_HEIGHT = 100


# 2 Pointer


def solve1(heights: List[int]) -> int:

    left, right = 0, len(heights) - 1
    left_max, right_max = 0, 0
    volume = 0

    while left < right:
        left_max, right_max = max(left_max, heights[left]), max(right_max, heights[right])

        if left_max <= right_max:
            volume += left_max - heights[left]
            left += 1
        else:
            volume += right_max - heights[right]
            right -= 1

    return volume


# Stack
def solve2(heights: List[int]) -> int:

    stack = []
    volume = 0

    for i in range(len(heights)):
        while stack and heights[i] > heights[stack[-1]]:
            top = stack.pop()

            if not len(stack):
                break

            distance = i - stack[-1] - 1
            waters = min(heights[i], heights[stack[-1]]) - heights[top]

            volume += distance * waters

        stack.append(i)

    return volume


if __name__ == "__main__":
    random.seed(17)

    for _ in range(TEST_CASES):

        n = random.randrange(2, N) + 2
        h = random.randrange(1, MAX_HEIGHT) + 1
        heights = [random.randrange(h) for _ in range(n)]
        print(solve1(heights), solve2(heights))
